### ISSUE TITLE
Camera calibration

### ISSUE THREAD

**Issue by adjelouah** [mcordts/cityscapesScripts#28]
Hello,

I am trying to compute the camera projection matrix from the intrinsic and extrinsic parameters following the description in: https://github.com/mcordts/cityscapesScripts/blob/master/docs/csCalibration.pdf
however the matrix I obtain is not correct and I suspect that I am misinterpreting some of the instruction.

Here is the snippet of code (based on opencv) I am using which should correspond exactly to the  provided description.

```
    // assuming all the values are read from the file
    
    // cos/sin for yaw, pitch, roll
    double c_y = cos(yaw);
    double s_y = sin(yaw);

    double c_p = cos(pitch);
    double s_p = sin(pitch);

    double c_r = cos(roll);
    double s_r = sin(roll); 
    
    // camera to vehicle rotation
    cv::Matx33d R(  c_y * c_p,  c_y * s_p * s_r - s_y*c_r,      c_y*s_p*c_r + s_y * s_r,
                    s_y * c_p,  s_y * s_p * s_r + c_y * c_r,    s_y*s_p*c_r - c_y * s_r, 
                    - s_p,      c_p * s_r,                      c_p * c_r);

    // camera rotation
    R = R.t();

    // camera t
    cv::Matx31d t = - ( R * cv::Matx31d(x,y,z) );

    // skipping steps to copy the values in the 3x4 Rt 

    // Intrinsic and final Projection matrix
    K = cv::Matx33d( fx,  0,     u0,
                   0,  fy,     v0,
                   0,  0,      1);

     cv::Matx33d intrinsic_rot(      0,  -1,     0,
                                    0,  0,      -1,
                                    1,  0,      0);

      K = K * intrinsic_rot;
      P = K * Rt
```
For example, given camera position (1.7, 0.1, 1.22) (in train/aachen)  projecting the 3d point (5.4, 0.1, 2.22) gives a 2d point at coordinate (u = 1052,  v = -192). This point far in front of the camera (4 meters) should be visible in the image.

Many thanks
Aziz

**Comment by jaypatravali** [comment]
Maybe just use disparity images for extracting world coordinates? Z= f*baseline/Disparity

**Comment by jaypatravali** [comment]
@adjelouah did you get a solution to this?

**Comment by adjelouah** [comment]
Hi,

Not yet. My issue was not with the depth value which I indeed obtain with the formula: Z = f*baseline/disp
The problem is the projection matrix. Using this depth values, the 3d points do not correctly project into the second camera. I think I am probably missing some transformations between the different systems of coordinates (car, world) but couldn't figure that from the documentation

For the moment I am using a different dataset and I won't be able to test again before a few weeks. The best would be to mark the issue as solved and I will just re open it when I come back to this dataset and provide a proper test scenario.

thanks,
Aziz

**Comment by jaypatravali** [comment]
@adjelouah 
I ask because I have been working on the image pixel->Camera->Vehicle World transformation. My task is to transform a pixel coordinate in the dataset images to a GPS coordinate in real world. I use the Vehicle JSON files which gives me the GPS Lat Lon and heading information.

What I observe from testing on multiple images is that in triangulating these points, the GPS measurements are not exactly captured where the image where taken. I have my own  dataset, where the same code gives desired measurements.

Not sure if this helps, but there is something definitely off with the Vehicles measurements given in the JSON files.

**Comment by mostajabi** [comment]
@adjelouah 
I have the same problem for rectifying stereo pairs, could you figure it out?

**Comment by jaypatravali** [comment]
@mostajabi 

TL;DR: there's no solution. My best guess is GPS stamps are noisy.

Going through some examples, it was quite clear from visual cues that the GPS tags were quite off from the actual position of the images.
For ex, in one of the dataset images, the car chassis covers a half of road sign for which the GPS location was 25m behind. Likewise, many such examples.

**Comment by mcordts** [comment]
Unfortunately, GPS can be noisy in inner-city traffic.
